rust/src/tensor.rs

```diff
@@ -0,0 +1,385 @@
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! Arrow Tensor Type
+use std::marker::PhantomData;
+use std::mem;
+
+use buffer::Buffer;
+use datatypes::{ArrowPrimitiveType, DataType};
+
+/// Computes the strides required assuming a row major memory layout
+fn compute_row_major_strides<T>(shape: &Vec<i64>) -> Vec<i64>
+where
+    T: ArrowPrimitiveType,
+{
+    let mut remaining_bytes = mem::size_of::<T>();
+    for i in shape {
+        remaining_bytes = remaining_bytes
+            .checked_mul(*i as usize)
+            .expect("Overflow occurred when computing row major strides.");
+    }
+
+    let mut strides = Vec::<i64>::new();
+    for i in shape {
+        remaining_bytes /= *i as usize;
+        strides.push(remaining_bytes as i64);
+    }
+    strides
+}
+
+/// Computes the strides required assuming a column major memory layout
+fn compute_column_major_strides<T>(shape: &Vec<i64>) -> Vec<i64>
+where
+    T: ArrowPrimitiveType,
+{
+    let mut remaining_bytes = mem::size_of::<T>();
+    let mut strides = Vec::<i64>::new();
+    for i in shape {
+        strides.push(remaining_bytes as i64);
+        remaining_bytes = remaining_bytes
+            .checked_mul(*i as usize)
+            .expect("Overflow occurred when computing column major strides.");
+    }
+    strides
+}
+
+/// Tensor of primitive types
+pub struct Tensor<'a, T>
+where
+    T: ArrowPrimitiveType,
+{
+    data_type: DataType,
+    buffer: Buffer,
+    shape: Option<Vec<i64>>,
+    strides: Option<Vec<i64>>,
+    names: Option<Vec<&'a str>>,
+    _marker: PhantomData<T>,
+}
+
+macro_rules! impl_tensor {
+    ($data_ty:path, $native_ty:ident) => {
+        impl<'a> Tensor<'a, $native_ty> {
+            /// Creates a new `Tensor`
+            pub fn new(
+                buffer: Buffer,
+                shape: Option<Vec<i64>>,
+                strides: Option<Vec<i64>>,
+                names: Option<Vec<&'a str>>,
+            ) -> Self {
+                match &shape {
+                    None => {
+                        assert_eq!(
+                            buffer.len(),
+                            mem::size_of::<$native_ty>(),
+                            "underlying buffer should only contain a single tensor element"
+                        );
+                        assert_eq!(None, strides);
+                        assert_eq!(None, names);
+                    }
+                    Some(ref s) => {
+                        strides
+                            .iter()
+                            .map(|i| {
+                                assert_eq!(s.len(), i.len(), "shape and stride dimensions differ")
+                            }).next();
```

Comment:
Is it better that we use `for_each()` instead of `map().next()` to reduce temporary `std::iter::Map`?

Reply:
I think you might be right about this one, I'll have to look into it when I get a chance...

Reply:
Thanks @kou